### DISCUSSION TITLE
Get rid of auto-currying, add syntax for currying/partial app.

### DISCUSSION THREAD

**Discussion by stylewarning** [coalton-lang/coalton#1071] [Ideas and Proposals]
"Auto-currying" is a very cool feature for some very slick and compact code. I love being able to write `(* 2)` as a doubling function, for example.

However, I find that in general, auto-currying is more trouble than the entry fee it paid:

- A lot of functions aren't useful as curried.
- Many functions are better curried when flipped.
- A typo, like not including an argument when one was intended, can lead to very difficult to debug type errors.
- We work hard to de-curry as much as possible.

As such, I'd propose we break an ML/Haskell tradition and get rid of it as a default, and instead use another syntax. Some ideas follow. In the following ideas, the spirit is what I'm aiming for, i.e., whether it's `_` or `$` or something else is open to discussion.

## Idea #1: Curry syntax

Suppose `f : t -> t -> t`. The syntax `[f x]` would represent the curried function `t -> t`.

## Idea #2: Same as #1, but with partial application

Suppose the same `f`. Then the above could be written `[f x _]`. The other argument could be `[f _ x]`. Multiple arguments could be `f == [f _1 _2]` and `flip f == [f _2 _1]`.

**Comment by sabjohnso** [comment]
My personal preference would be to keep auto-currying.  However, I think a syntax for the abbreviated definition of anonymous functions would be valuable regardless, and I do recognize there is merit to having an alternate syntax for partial application as it would communicate more information to readers of the code.

Below are some related ideas from other languages.  In particular, Racket's  curly-fn library essentially combines Idea1 and Idea2.

### SRFI 26
https://srfi.schemers.org/srfi-26/srfi-26.html
Scheme's SRFI 26 is conceptually similar to the proposed syntax but
also has some striking differences. For partial application the syntax
would be similar to the propopsed syntax:

```scheme
(cut f x <>)
```
```scheme
(fn (y) (f x y)
```

or for right partial appliction

```scheme
(cut f <> x)
```

```scheme
(fn (y) (f y x))
```

Note for a ternary function `h : :t -> :t -> :t -> :t`,
```scheme
(cut h x <> <>)
```
would be equivalent to
```scheme
(fn (y z) (h x y z))
```
and would have the type `:t -> :t -> :t`, not `:t -> :t`. Namely, the first and second
lozenges do not refer to the same argument. Rather, the lozenges just
represents holes that are left open. As a result, `cut` cannot be used
to reorder the arguments to `f`. `Cut` has some additional
limitations/draw backs:

```scheme
(cut g (f x <>) <>)
```
would be equivalent to
```scheme
(fn (y) (g (f x <>) y)
```

and would likely fail to compile because `(f x <>)` is read as the
function `f` applied to the arguments `x` and `<>`, where `<>` would
be undefined. However, there is a more horrifying scenario: `<>` is
actually define in the lexical environment. In that scenario, `g` is
partially applied to the full application of `f` to `x` and `<>` as
bound in the lexical environment. Probably not the intent of the code
and definitely confusing to readers.

The `cut` forms can be nested but my may still be difficult to read
```scheme
(cut g (cut f x <>) <>)
```
would be the partial application of g to the partial application of
`f` to `x` or
```scheme
(fn (z) 
  (fn (y)
    (g (f x y) z)))
```

Finally  `(cut f x y)`, where `x` and `y` are defined in the lexical
environment, defines a thunk.

### Clojure's Anonymous function syntax
https://clojure.org/guides/learn/functions#_anonymous_function_syntax
Clojure has a syntax for the abbreviated definition of of anonymous
functions, which is similar to the proposed syntax.

The partial application of `f` would be as follows:
``` clojure
  #(f x %)
```
and swapping the argument in `f` would be as follows:
``` clojure
 #(f %2 %1)
```
Unlike `cut`, Clojure's syntax allows `%` to be used in compound
subforms, not just in the top of the `cut` form:
```clojure
#(g (f %) ) %)
```
would be equivalent to

```clojure
(fn (x) (g (fn (y) (f x y)) x)
```

Clojure does not allow nesting of abbreviated function definitions:
```clojure
#(g #(f x %) %)

```
is a syntax error. This limitation means it cannot replace 
`(g (f x))`,  `fn` can always solve that problem. 


### Racket
https://github.com/lexi-lambda/racket-curly-fn
Racket has a library implementing  a metalanguage for abbreviated
definitions of anonymous functions that is similar to Clojure's but
using braces instead of parentheses to avoid a conflict with the
syntax for literal vectors.

The partial application of `f` would be as follows:

``` racket
#{f x %}
```
and swapping the arguments would be
```racket
#{f %2 %1}
```

It also supports the curried version like Idea 1, which Clojure's
syntax is not capable of. Namely `#{f x}` is equivalent to `#{f x %}`,
and is a suitable replacement for `(g (f x))`: `#{g #{f x}}`.

This is essentially the same as Idea1 and Idea2 combined, with
different delimiters to open and close the form.


### Mathematica
https://www.wolfram.com/language/elementary-introduction/2nd-ed/26-pure-anonymous-functions.html
Mathematica also has a similar notion where 
``` mathematica
F[x, #]&
```

is the partial application of `F` and

``` mathematica
F[#2, #1]&
```

**Comment by stylewarning** [comment]
Inspired by [Joseph Garvin](https://x.com/joseph_h_garvin) on Twitter, I started thinking about this question again and have been prototyping out a new proposal. It's basically as follows.

1. Functions take a fixed number of inputs and produce a fixed number of outputs.

This is pretty self-explanatory. There's no "arity-polymorphism". A function that expects 5 arguments must be supplied 5 arguments. A function that produces 2 values must be called in a place where 2 values can be received. (If it is unknown how many return values a function has, it is assumed to be one.)

Corollary: There is no auto-currying. Any desired currying has to be done manually.

2. Functions that take multiple arguments are typed differently than functions that return functions.

In Coalton today, we have (for example) the following:

```
(declare f (String -> String -> String))
(define (f a b)
  ...)
```

This doesn't truly take 2 arguments—it takes one argument `a` and produces a new function as a result—but we all get what it means when we say it takes 2 arguments. In the proposed change, this function would instead be written

```
(declare f (String * String -> String)
(define (f a b)
  ...)
```

Unlike other MLs, the `*` here is attached to the `->`. It does *not* represent a tuple type.

Similarly, we could write a function which returns two values, like Common Lisp's `gethash`:

```
(declare gethash (HashTable :k :v * :k -> :v * Boolean))
```

I don't advise we have this in Coalton, but you get the idea. The two return values would be passed as Common Lisp `cl:values`. In fact, Coalton would also have a new special operator `coalton:values` that serves the same exact purpose.

3. Multiple values must be received by a binding form or returned. There is no implicit "first value" handling.

If we have the `gethash` above, we must call against a binding form, or return it from the function. Multiple values can only exist in those two places. As such, I propose there are 2 places this can happen: the short-`let` or the normal `let`:

```
(let (((values v found?) (gethash tbl x)))
  ...)

;; or

(let (values v found?) = (gethash tbl x))

;; but not

(let v = (gethash tbl x))
```

4. There is no values polymorphism

Just like there's no arity-polymorphism (being able to abstract over the arity of a function), the same is true for the number of returned values. As such, it's not possible to write a "universal" identity function, for example. Values must either be returned or received explicitly. If it's ambiguous how many values a function could return, it's assumed to be 1, as in the following:

```
(define (f ambiguous-returns x)
  (ambiguous-returns x))
```

The function `ambiguous-returns` in principle could return any number of arguments, but since it's not constrained and since there's no return-value polymorphism, it's assumed to be `:s -> :t`.

5. We remove `Tuple` unboxing, for now.

Unboxing `Tuple` was added because we didn't support multiple values. But it adds tremendous implementation complexity. Just keep Tuple boxed, at least for now, and use multiple values. We can revisit a more elegant solution to unboxing at a later point, perhaps via @shirok's ideas below.

6. This opens up the possibility for alternative calling conventions in the future.

Under the hood, `a * b * c ... -> d * e * f ...` basically gets parsed as `(CoaltonFunctionInputs a b c) -> (CoaltonFunctionOutputs d e f)`. In principle, we could have `(CFunctionInputs ...) -> (CFunctionOutputs ...)` or other special calling/return conventions.

7. The concept of under- or over-application disappears.

We get real arity errors now. Hindley-Milner type errors earn a little more "locality" since bad currying can't stray too far.

8. Keyword arguments are trivially supported.

There's a PR up adding keyword arguments to Coalton. They're the best I could do, but they're kind of bizarre:

```
(declare f (:s -> :t -> (&key (:foo Integer)) -> :u))
```

When you apply :t, you have one immediate opportunity to also supply the keyword arguments, or decline. If you curry :t without keywords, the keywords are declined. This just makes an unnatural relationship between both sides of the arrow.

In this proposal, we have a natural place for keyword arguments: before the arrow.

```
(declare f (:s * :t &key (:foo Integer) -> :u))
```

Keywords are supplied with the fixed inputs, or they're not. Straightforward.

***

I have a half-baked implementation of this all, and the funny thing is, *99% of the actual executable code remains unbroken*! The only things that break are:

- The function declarations: `a -> b -> c` usually has to be rewritten to `a * b -> c`.
- A few "very high order" partial application stuff, that's solved by eta-expansion.

However, there are much more serious issues to grapple with: inference. What is the type of this function?

```
(defun (a f x)
  (f x))
```

In my prototype system, it's `(declare a ((:s -> :t) * :s -> :t))`, which means `f` can't be a multiple-value-returning function despite it being legal in principle.

I haven't yet prototyped out any nice lambda shorthands, some of which @sabjohnso nicely outlined.

I'm not 100% convinced this is the right move for Coalton, but it does make certain optimizations a lot more straightforward, and it pulls us away from the sort of subtle lie we tell ourselves that we're basically like Haskell (we're very much not).

I'll update more as I progress, until I'm able to make a PR.

**Reply by c-kloukinas** [reply]
> Unlike other MLs, the * here is attached to the ->. It does not represent a tuple type.

What type would it represent, if not a tuple?

And what would be the advantage of that other type compared to a tuple?

**Reply by stylewarning** [reply]
> > Unlike other MLs, the * here is attached to the ->. It does not represent a tuple type.
> 
> 
> 
> What type would it represent, if not a tuple?
> 
> 
> 
> And what would be the advantage of that other type compared to a tuple?

A tuple is conventionally a heap-allocated collection of pointers with a type tag. The proposal above tries to avoid that by natively representing functions with different input-output arities. You can't construct a `Integer * Integer` as a value.

**Comment by shirok** [comment]
Existence of auto-curry affects API design as well.   Suppose you have an operation on some object.  If the operation takes a parameter, languages with auto-currying tend to make `(op param object)`, while languages without auto-currying tend to make `(op object param)`.

I've been in the latter camp for long time and tend to make things in that way---e.g. many HashMap API takes a hashmap as the first arguemnt.  On the other hand, Haskell's `Ddata.HashMap` tend to take a hashmap as the last argument.  With that regard, Coalton's standard library has mixed flavors now.

Somehow I find myself using much less curried calls in Coalton than in Haskell.  I'd be happly to have some shorthand notation.  By the way, Paul Graham's Arc uses `[f x _]` for `(fn (a) (f x a))` etc.

**Comment by shirok** [comment]
By the way, regarding Tuples---I dreamed to represent Tuples with multiple values by default, i.e. the continuation of Tuple-yielding expression takes multiple arguments.   It's the continuation's responsibility to box them into a Tuple struct whenver needed.    This would be easier if the compiler is written in CPS, but may not be trivial in tree-walking style.
